interpreter/attributes_test.go

```diff
@@ -0,0 +1,420 @@
+	data := map[string]interface{}{
+		"a": map[int]interface{}{
+			-1: []int32{2, 42},
+		},
+		"b": 1,
+	}
+	vars, _ := NewActivation(data)
+
+	// obj.a[-1][b]
+	op := &evalConst{
+		id:  1,
+		val: reg.NativeToValue(data),
+	}
+	attr := attrs.RelativeAttribute(1, op)
+	qualA, _ := attrs.NewQualifier(nil, 2, "a")
+	qualNeg1, _ := attrs.NewQualifier(nil, 3, int64(-1))
+	attr.AddQualifier(qualA)
+	attr.AddQualifier(qualNeg1)
+	attr.AddQualifier(attrs.AbsoluteAttribute(4, "b"))
+	out, err := attr.Resolve(vars)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if out != types.Int(42) {
+		t.Errorf("Got %v (%T), wanted 42", out, out)
+	}
+}
+
+func TestAttributes_RelativeAttr_OneOf(t *testing.T) {
+	reg := types.NewRegistry()
+	pkg := packages.NewPackage("acme.ns")
+	attrs := NewAttributeFactory(pkg, reg, reg)
+	data := map[string]interface{}{
+		"a": map[int]interface{}{
+			-1: []int32{2, 42},
+		},
+		"acme.b": 1,
+	}
+	vars, _ := NewActivation(data)
+
+	// obj.a[-1][b]
+	op := &evalConst{
+		id:  1,
+		val: reg.NativeToValue(data),
+	}
+	attr := attrs.RelativeAttribute(1, op)
+	qualA, _ := attrs.NewQualifier(nil, 2, "a")
+	qualNeg1, _ := attrs.NewQualifier(nil, 3, int64(-1))
+	attr.AddQualifier(qualA)
+	attr.AddQualifier(qualNeg1)
+	attr.AddQualifier(attrs.OneofAttribute(4, "b"))
+	out, err := attr.Resolve(vars)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if out != types.Int(42) {
+		t.Errorf("Got %v (%T), wanted 42", out, out)
+	}
+}
+
+func TestAttributes_RelativeAttr_Conditional(t *testing.T) {
+	reg := types.NewRegistry()
+	attrs := NewAttributeFactory(packages.DefaultPackage, reg, reg)
+	data := map[string]interface{}{
+		"a": map[int]interface{}{
+			-1: []int32{2, 42},
+		},
+		"b": []int{0, 1},
+		"c": []interface{}{1, 0},
+	}
+	vars, _ := NewActivation(data)
+
+	// obj[a][-1][(cond ? b : c)[0]]
+	cond := &evalConst{
+		id:  2,
+		val: types.False,
+	}
+	condAttr := attrs.ConditionalAttribute(4, cond,
+		attrs.AbsoluteAttribute(5, "b"),
+		attrs.AbsoluteAttribute(6, "c"))
+	qual0, _ := attrs.NewQualifier(nil, 7, 0)
+	condAttr.AddQualifier(qual0)
+
+	obj := &evalConst{
+		id:  1,
+		val: reg.NativeToValue(data),
+	}
+	attr := attrs.RelativeAttribute(1, obj)
+	qualA, _ := attrs.NewQualifier(nil, 2, "a")
+	qualNeg1, _ := attrs.NewQualifier(nil, 3, int64(-1))
+	attr.AddQualifier(qualA)
+	attr.AddQualifier(qualNeg1)
+	attr.AddQualifier(condAttr)
+	out, err := attr.Resolve(vars)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if out != types.Int(42) {
+		t.Errorf("Got %v (%T), wanted 42", out, out)
+	}
+}
+
+func TestAttributes_RelativeAttr_Relative(t *testing.T) {
+	pkg := packages.NewPackage("acme.ns")
+	reg := types.NewRegistry()
+	attrs := NewAttributeFactory(pkg, reg, reg)
+	data := map[string]interface{}{
+		"a": map[int]interface{}{
+			-1: map[string]interface{}{
+				"first":  uint(1),
+				"second": uint(2),
+				"third":  uint(3),
+			},
+		},
+		"b": uint32(2),
+	}
+	vars, _ := NewActivation(data)
+
+	// a[-1][{1u: "first", 2u: "second", 3u: "third"}[b]]
```

Comment:
The comment doesn't match what's evaluated - it should have "obj" or `{"a": , "b": , "c": }` in there somewhere.